### ISSUE TITLE
2.3 --> 5.1 Migration : ip type field doesn't work for 2.3 migrated indexes and ipv6 data

### ISSUE THREAD

**Issue by r39132** [elastic/elasticsearch#23126]
**Elasticsearch version**: 5.1 (AWS hosted ES)

**Plugins installed**: []

**JVM version**: 1.7

**OS version**: N/A

**Description of the problem including expected versus actual behavior**:

I migrated from 2.3 to 5.1 following the steps listed [here](http://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-version-migration.html). 

Essentially, I did this by snapshotting the 2.3 indexes to S3 and then restoring them to the 5.1 cluster. All seemed fine until I added an `ip` type field to my indexes, which consisted of some indexes ported from 2.3 and some created anew on the 5.1 cluster. FYI, my indexes are day-wise (e.g. message_20170101, message_20170102, message_20170103 ). 

When I tried to insert an ipv6 value in to the fields in the new indexes, all worked fine. However, when I tried to insert the same value in the 2.3-migrated indexes, I got an exception `failed to parse ip [2607:f8b0:400e:c00::22b], not a valid ipv4 address (4 dots)`.

Here's an example mapping for one of the 2.3-migrated indexes that did not work. The 2.3 index contained 4 fields : `color, make, sold, and price`. After restoring the index on 5.1 and after adding the `my_ip` field, the mapping looks as shown below:

```
deploy@es-proxy-00:~$ curl -XGET 'localhost:80/cars/transactions/_mapping'
{
  "cars": {
    "mappings": {
      "transactions": {
        "properties": {
          "color": {
            "type": "string"
          },
          "make": {
            "type": "string"
          },
          "my_ip": {
            "type": "ip"
          },
          "price": {
            "type": "long"
          },
          "sold": {
            "type": "date",
            "format": "strict_date_optional_time||epoch_millis"
          }
        }
      }
    }
  }
}
```

Here's what happened when I tried to insert an ipv6 value into that field into one of the docs.

```
deploy@es-proxy-00:~$ cat test_ip_a.sh 
curl -XPOST "localhost:80/cars/transactions/AVlxOqfz-i1bJQCCyerj" -d'
{
      "my_ip": "2607:f8b0:400e:c00::22b"
}'

deploy@es-proxy-00:~$ ./test_ip_a.sh 
{
  "error": {
    "root_cause": [
      {
        "type": "mapper_parsing_exception",
        "reason": "failed to parse [my_ip]"
      }
    ],
    "type": "mapper_parsing_exception",
    "reason": "failed to parse [my_ip]",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "failed to parse ip [2607:f8b0:400e:c00::22b], not a valid ipv4 address (4 dots)"
    }
  },
  "status": 400
}
```

**Steps to reproduce**:
 1. Migrate from 2.3 to 5.1 : snapshot 2.3 to S3 and restore from S3 to the 5.1 cluster
 2. Create a new index on 5.1 -- call it new-foo
 3. Add an ip field to new-foo and the 2.3 migrated indexes
 4. Insert an ipv6 field value (`e.g. 2607:f8b0:400e:c00::22b`) into all of these indexes
 5. It will fail for the 2.3-migrated indexes and work for the new-foo index

**Provide logs (if relevant)**:

N/A

**Describe the feature**:

N/A

**Comment by clintongormley** [comment]
> However, when I tried to insert the same value in the 2.3-migrated indexes, I got an exception failed to parse ip [2607:f8b0:400e:c00::22b], not a valid ipv4 address (4 dots).

Yes - the new data structure is only available on indices created in 5.0 - the migrated indices still use the old datastructure. You should either use a new index, or use reindex to migrate your old index to a new index

**Comment by r39132** [comment]
@clintongormley thx for the reply. In our case, for space saving, we don't keep the source, so we are not able to reindex.
